src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsReader.cs

```diff
@@ -0,0 +1,227 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers.Binary;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Data.SqlClientX.IO
+{
+    /// <summary>
+    /// This class provides helper methods for reading bytes using <see cref="TdsStream"/>
+    /// It extends <see cref="TdsBufferManager"/> that manages allocations of bytes buffer for better memory management.
+    /// </summary>
+    internal class TdsReader : TdsBufferManager
+    {
+        private readonly TdsStream _tdsStream;
+
+        /// <summary>
+        /// Instantiate TdsReader with <see cref="TdsStream" />
+        /// </summary>
+        /// <param name="stream">Tds Stream instance to work with.</param>
+        public TdsReader(TdsStream stream)
+        {
+            _tdsStream = stream;
+        }
+
+        #region Public APIs
+
+        /// <summary>
+        /// Reads byte from TDS stream asynchronously.
+        /// </summary>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token.</param>
+        /// <returns>Async <see cref="ValueTask"/> that returns a <see cref="byte"/> value.</returns>
+        public async ValueTask<byte> ReadByteAsync(bool isAsync, CancellationToken ct)
+        {// Throw operation canceled exception before write.
+            ct.ThrowIfCancellationRequested();
+
+            return isAsync ?
+                await _tdsStream.ReadByteAsync(isAsync, ct).ConfigureAwait(false)
+                : (byte)_tdsStream.ReadByte(); // UNSAFE - TODO Expose ReadByte from TdsStream that returns 'byte'.
+        }
+        /// <summary>
+        /// Reads next char from TDS stream asynchronously.
+        /// </summary>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token.</param>
+        /// <returns>Async <see cref="ValueTask"/> that returns a <see cref="char"/> value.</returns>
+        public ValueTask<char> ReadCharAsync(bool isAsync, CancellationToken ct)
+            => DoWithRentedBuffer(sizeof(char), async buffer =>
```

Comment:
You might want to benchmark, how much this extra functor indirection affects performance, besides readability. I would personally not do it but use the pool directly in here

Reply:
I think that even more important point, is the use of asynchronous lambdas or functions; each aync function adds an additional state machine, which can add up to a lot of overhead. So the general tendency to refactor common code out to smaller functions doesn't work well for async.

As a very general guidance, I'd recommend going over where async is used, and trying to minimize the instances of that, generally concentrating it in a few few big functions rather than spreading it all around in a lot of small helpers.

Reply:
Simplified